### PR TITLE
Add `ConvertTo-CliXml` and `ConvertFrom-CliXml` cmdlets

### LINKED ISSUE #3898: Export-CliXml shouldn't require writing to a file
The `Export-CliXml` command should offer a parameter set that emits a string, instead of writing the result to a file. Right now, it only supports outputting CLIXML to a file via the `-Path` or `-LiteralPath` parameters.

Cheers,
[Trevor Sullivan](https://trevorsullivan.net)

### PR DESCRIPTION
**Pull request by ArmaanMcleod**
# PR Summary

Fixes #3898

Add `ConvertTo-CliXml` and `ConvertFrom-CliXml` cmdlets to work with CliXml objects in memory without needing to read/write to file system.

### Parameter sets

```pwsh
ConvertTo-CliXml [-InputObject] <psobject> [-Depth <int>] [<CommonParameters>]

ConvertFrom-CliXml [-InputObject] <string> [<CommonParameters>]
```

### Usage

```pwsh
> $cliXmlString = [pscustomobject]@{'hello' = 1} | ConvertTo-CliXML
> $cliXmlString
<Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell/2004/04">
  <Obj RefId="0">
    <TN RefId="0">
      <T>System.Management.Automation.PSCustomObject</T>
      <T>System.Object</T>
    </TN>
    <MS>
      <I32 N="hello">1</I32>
    </MS>
  </Obj>
</Objs>
> $cliXmlString | ConvertFrom-CliXml

hello
-----
    1
```

Also included `internal static string Serialize(IList<object> source, int depth, bool enumerate)` in `PSSerializer` API. This was to be able to enumerate and serialize objects one at a time instead of writing one top level object. 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/10773
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs [hunk 1]
```diff
@@ -632,6 +632,88 @@ private void CleanUp()
         #endregion IDisposable Members
     }
 
+    /// <summary>
+    /// Implements ConvertTo-CliXml command.
+    /// </summary>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2243:AttributeStringLiteralsShouldParseCorrectly")]
+    [Cmdlet(VerbsData.ConvertTo, "CliXml", HelpUri = "")]
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:639]: /cc @sdwheeler Please create HelpUri-s for new cmdlwts.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1 [hunk 2]
```diff
@@ -340,4 +346,200 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
...
+            $ps.BeginInvoke()
+            $ps.Stop()
+            $ps.InvocationStateInfo.State | Should -BeExactly "Stopped"
+        }
+
+        It "Should serialize integers correctly using ValueFromPipeline" {
+            $testObject = 1,2,3
+            $content = $testObject | ConvertTo-CliXml
+            $testObject | Export-CliXml -Path $testfile
+            $content | Should -Be (Get-Content -Path $testfile -Raw)
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:399]: Maybe https://pester.dev/docs/v4/commands/Should#filecontentmatchexactly

**Review by iSazonov — APPROVED** [review]
LGTM with minor comments

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1 (hunk 2 above)
Comment (ArmaanMcleod) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:399]: I think we can actually use https://pester.dev/docs/v4/commands/Should#filecontentmatchmultiline, which is better for matching multi line files.

Comment (ArmaanMcleod) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1]: Fixed in fdb200380bc9d4c3bdf2da20e98c3e3bfe90617a.

**Review by JamesWTruher — APPROVED** [review]
Really nice, just some comments about the tests

File: test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1 [hunk 3]
```diff
@@ -181,6 +185,96 @@ Describe "CliXml test" -Tags "CI" {
...
+
+        It "Should import PSCredential" {
+            $UserName = "Foo"
+            $pass = ConvertTo-SecureString (New-RandomHexString) -AsPlainText -Force
+            $cred =  [PSCredential]::new($UserName, $pass)
+
+            $content = $cred | ConvertTo-CliXml
+            $cred = ConvertFrom-CliXml -InputObject $content
+            $cred.UserName | Should -BeExactly "Foo"
+            $cred.Password | Should -BeOfType System.Security.SecureString
```
[the remark this answers is not in the document]
Comment (ArmaanMcleod) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1]: Thanks @JamesWTruher. I have updated this in 64a1e76f9.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1 [hunk 4]
```diff
@@ -340,4 +346,200 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
 		$cmd.Xml = $xml
 		$cmd.Xml | Should -Be $xml
 	}
+
+    Context "ConvertTo-CliXml & ConvertFrom-CliXml" {
+
+        It "Getting cmdlet info should work" {
+            $content = Get-Command export* -Type Cmdlet | Select-Object -First 3 | ConvertTo-CliXml
```
[the remark this answers is not in the document]
Comment (ArmaanMcleod) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1]: Thanks @JamesWTruher. I have updated this to only include 2 commands in 64a1e76f9.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1 [hunk 5]
```diff
@@ -340,4 +346,200 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
...
+            $ps.BeginInvoke()
+            $ps.Stop()
+            $ps.InvocationStateInfo.State | Should -BeExactly "Stopped"
+        }
+
+        It "Should serialize integers correctly using ValueFromPipeline" {
+            $testObject = 1,2,3
+            $content = $testObject | ConvertTo-CliXml
+            $testObject | Export-CliXml -Path $testfile
+            $testfile | Should -FileContentMatchMultiline ([regex]::Escape($content))
```
[the remark this answers is not in the document]
Comment (ArmaanMcleod) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1]: Thanks @JamesWTruher. I have removed the pester match multi line check and used your suggestion in 64a1e76f9

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1 [hunk 6]
```diff
@@ -340,4 +346,200 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
...
+            $out = ConvertFrom-CliXml -InputObject $content
+            $out -is [array] | Should -BeTrue
+            $out.Count | Should -Be 3
+            $out[0] | Should -Be 1
```
[the remark this answers is not in the document]
Comment (ArmaanMcleod) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:403]: Thanks @JamesWTruher. I have updated this in 64a1e76f9

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1 [hunk 7]
```diff
@@ -340,4 +346,200 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
...
+            $deserialized_one = ConvertFrom-CliXml -InputObject $content
+            $deserialized_one.value | Should -Be 1
+            $deserialized_one.two | Should -Not -BeNullOrEmpty
+            $deserialized_one.two.value | Should -Be 2
+            $deserialized_one.two.three | Should -Not -BeNullOrEmpty
+            $deserialized_one.two.three.num | Should -Be 3
+            $deserialized_one.two.three.four | Should -Not -BeNullOrEmpty
+            $deserialized_one.two.three.four.num | Should -BeNullOrEmpty
+        }
+
```
[the remark this answers is not in the document]
Comment (ArmaanMcleod) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:438]: Thanks @JamesWTruher. I have included another test in 64a1e76f9 to check this :)

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1 [hunk 8]
```diff
@@ -340,4 +346,200 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
...
+            $testfile | Should -FileContentMatchMultiline ([regex]::Escape($content))
+            $out = ConvertFrom-CliXml -InputObject $content
+            $out -is [pscustomobject] | Should -BeTrue
+            $out.a | Should -Be 1
+            $out.b | Should -Be 2
+            $out.c | Should -Be 3
+            $out.d -is [pscustomobject] | Should -BeTrue
+            $out.d.e | Should -Be 4
+        }
+
```
[the remark this answers is not in the document]
Comment (ArmaanMcleod) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:535]: Yeah interesting for sure, atleast this makes the tests simpler :)

**Comment by ArmaanMcleod** [comment]
@JamesWTruher Thanks for the PR feedback, I have made the changes to the tests. Let me know if you want anything else changed 🙂

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by iSazonov** [comment]
@sdwheeler We need HelpURI for the new cmdlets. Please create.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs (hunk 1 above)
Comment (sdwheeler) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:639]:
```suggestion
    [Cmdlet(VerbsData.ConvertTo, "CliXml", HelpUri = "https://go.microsoft.com/fwlink/?LinkID=2280866")]
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs [hunk 9]
```diff
@@ -632,6 +632,88 @@ private void CleanUp()
...
+        }
+
+        #endregion Overrides
+    }
+
+    /// <summary>
+    /// Implements ConvertFrom-CliXml command.
+    /// </summary>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2243:AttributeStringLiteralsShouldParseCorrectly")]
+    [Cmdlet(VerbsData.ConvertFrom, "CliXml", HelpUri = "")]
```
Comment (sdwheeler) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:691]:
```suggestion
    [Cmdlet(VerbsData.ConvertFrom, "CliXml", HelpUri = "https://go.microsoft.com/fwlink/?LinkID=2280770")]
```

**Review by sdwheeler — CHANGES_REQUESTED** [review]
Please commit the HelpUri changes.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs [hunk 10]
```diff
@@ -632,6 +632,88 @@ private void CleanUp()
         #endregion IDisposable Members
     }
 
+    /// <summary>
+    /// Implements ConvertTo-CliXml command.
+    /// </summary>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2243:AttributeStringLiteralsShouldParseCorrectly")]
```
Comment (ArmaanMcleod) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:638]: We can probably remove these suppressions after fixing HelpUri

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs [hunk 11]
```diff
@@ -632,6 +632,88 @@ private void CleanUp()
...
+            WriteObject(PSSerializer.Serialize(_inputObjectBuffer, Depth, enumerate: true));
+        }
+
+        #endregion Overrides
+    }
+
+    /// <summary>
+    /// Implements ConvertFrom-CliXml command.
+    /// </summary>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2243:AttributeStringLiteralsShouldParseCorrectly")]
```
Comment (ArmaanMcleod) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:690]: And this one as well

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs (hunk 10 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:638]: You can try :-)

Comment (ArmaanMcleod) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:638]:
```suggestion
```

Comment (ArmaanMcleod) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: Done

**Comment by iSazonov** [comment]
@ArmaanMcleod Thanks for your contribution!

**Pull request merged**